docs/arcadia-users-group/20240206-intro-to-formatting-and-linting/lesson.md

```diff
@@ -0,0 +1,425 @@
+def calculate_roots(quadratic_coeff, linear_coeff, constant_coeff):
+    """
+    Calculate the real roots of a second order polynomial of the form
+    a * x^2 + b * x + c = 0.
+
+    Args:
+        quadratic_coeff: The coefficient of the quadratic term, or `a` in the above equation.
+        linear_coeff: The coefficient of the linear term, or `b` in the above equation.
+        constant_coeff: The constant term, or `c` in the above equation.
+
+    Returns: 
+        A tuple containing the two real roots of the polynomial.
+    """
+    discriminant_square_root = np.sqrt(
+        linear_coeff ** 2 - 4 * quadratic_coeff * constant_coeff
+    )
+    root_1 = (-linear_coeff + discriminant_square_root) / (2 * quadratic_coeff)
+    root_2 = (-linear_coeff - discriminant_square_root) / (2 * quadratic_coeff)
+    return root_1, root_2
+```
+This version of the function is both easy to read and easy to understand; the docstring explains exactly what the function does, what its inputs are, and what its outputs are, and the implementation is transparent and easy to follow thanks to the descriptive variable names.
+
+
+## Linting
+"Linting" refers to the analysis of source code to detect syntax errors, potential bugs, and violations of style conventions. Linters are similar to formatters in the sense that they analyze source code and are used to ensure that code adheres to certain standards, but unlike formatters, they typically do not modify the code itself. Instead, linters typically flag issues (or "lint errors") for the developer to fix.
+
+Linters are particularly important for interpreted languages like Python that are dynamically typed and interpreted at runtime, because many errors that would be caught by the compiler in a statically typed language (like C or Java) are not caught until the code is actually run, by which time it may be difficult or time-consuming the fix the error. Linters can catch these kinds of errors at an early stage when they are easy for the developer to fix.
+
+For example, consider the following code:
+```python
+def show_full_name(first_name, last_name):
+    full_name = first_name + last_nam
+    print(f'Your full name is {full_name}!')
+```
+Note that there is a typo in the function: the variable `last_name` is misspelled as `last_nam`. Unfortunately, because Python is interpreted, this mistake will go unnoticed until the code is run and the `show_full_name` function is called. At this point, an error will occur (and the Python process will crash) because the variable `last_nam` is not defined. A linter, by analyzing the source code, can detect this kind of error before the code is ever run (much less shared with others or released) and flag it for the developer to fix.
+
+In Python, one popular linting tool is called `ruff`. Let's try it out. First, install `ruff` in our conda environment:
+```bash
+mamba install -y ruff
+```
+Now, copy the function above into `main.py` and run `ruff` from the command line:
```

Comment:
so you'll make sure learners are on the lesson and scrolling along with you, or post this into the slack chat so that everyone can have a copy?

Reply:
Good question! I can have it ready to copy-paste into slack and the zoom chat.